File: frontend/src/api/crossLayerIntelligence.ts
import { apiClient } from '@/api/client';

export interface InsightEvidence {
  layer: string;
  signal: string;
  value: unknown;
}

export interface CrossLayerInsight {
  insight_id: string;
  insight_type: string;
  severity: 'critical' | 'warning' | 'info' | 'opportunity';
  title: string;
  description: string;
  evidence: InsightEvidence[];
  recommendation: string;
  confidence: number;
  estimated_impact: string;
}

export interface IntelligenceSummary {
  total_insights: number;
  by_type: Record<string, number>;
  by_severity: Record<string, number>;
  top_critical: CrossLayerInsight[];
  computed_at: string;
}

export const crossLayerIntelligenceApi = {
  getBuildingIntelligence: async (buildingId: string): Promise<CrossLayerInsight[]> => {
    const response = await apiClient.get<CrossLayerInsight[]>(`/buildings/${buildingId}/intelligence`);
    return response.data;
  },

  getPortfolioIntelligence: async (): Promise<CrossLayerInsight[]> => {
    const response = await apiClient.get<CrossLayerInsight[]>('/portfolio/intelligence');
    return response.data;
  },

  getIntelligenceSummary: async (buildingId?: string): Promise<IntelligenceSummary> => {
    const params = buildingId ? { building_id: buildingId } : {};
    const response = await apiClient.get<IntelligenceSummary>('/intelligence/summary', { params });
    return response.data;
  },
};

File: backend/app/schemas/cross_layer_intelligence.py
"""Pydantic v2 schemas for the Cross-Layer Intelligence Engine."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict


class InsightEvidence(BaseModel):
    """Single evidence signal from one layer."""

    layer: str
    signal: str
    value: Any = None

    model_config = ConfigDict(from_attributes=True)


class CrossLayerInsightRead(BaseModel):
    """A cross-layer insight discovered by correlating multiple layers."""

    insight_id: str
    insight_type: str
    severity: str  # critical / warning / info / opportunity
    title: str
    description: str
    evidence: list[InsightEvidence] = []
    recommendation: str
    confidence: float  # 0-1
    estimated_impact: str

    model_config = ConfigDict(from_attributes=True)


class PortfolioInsightRead(BaseModel):
    """Portfolio-level insight (same shape, distinct type for clarity)."""

    insight_id: str
    insight_type: str
    severity: str
    title: str
    description: str
    evidence: list[InsightEvidence] = []
    recommendation: str
    confidence: float
    estimated_impact: str

    model_config = ConfigDict(from_attributes=True)


class IntelligenceSummaryRead(BaseModel):
    """Aggregated intelligence summary."""

    total_insights: int
    by_type: dict[str, int] = {}
    by_severity: dict[str, int] = {}
    top_critical: list[CrossLayerInsightRead] = []
    computed_at: str

    model_config = ConfigDict(from_attributes=True)
